util/fibers/listener_interface.cc

```diff
@@ -31,39 +31,68 @@ auto native_handle(const Connection& conn) {
 
 // We set intrusive::constant_time_size<true> because we use ListType::size()
 // for introspection purposes.
-using ListType = intrusive::list<Connection, Connection::member_hook_t,
-                                 intrusive::constant_time_size<true>>;
+using ListType =
+    intrusive::list<Connection, Connection::member_hook_t, intrusive::constant_time_size<true>>;
 
 }  // namespace
 
+thread_local unordered_map<ListenerInterface*, ListenerInterface::TLConnList*>
+    ListenerInterface::conn_list;
+
 struct ListenerInterface::TLConnList {
   ListType list;
   fb2::CondVarAny empty_cv;
 
   void Link(Connection* c) {
     DCHECK(!c->hook_.is_linked());
-    c->DEBUG_connlist_pthread_id = pthread_self();
+    c->DEBUG_proactor = ProactorBase::me();
     list.push_front(*c);
     DVLOG(3) << "List size " << list.size();
   }
 
-  void Unlink(Connection* c) {
+  void Unlink(Connection* c, ListenerInterface* l) {
     CHECK(c->hook_.is_linked());
 
-    if (c->DEBUG_connlist_pthread_id != pthread_self()) {
-      LOG(FATAL) << "inconsistent connection thread id " << c->DEBUG_connlist_pthread_id << " vs "
-                 << pthread_self();
+    auto it = ListType::s_iterator_to(*c);
+
+    {
+      util::ProactorBase *left = nullptr, *right = nullptr;
+      if (auto lit = it; it != list.begin())
+        left = (--lit)->DEBUG_proactor;
+      if (auto rit = it; ++rit != list.end())
+        right = rit->DEBUG_proactor;
```

Comment:
> I think we should check it regardless before and log error ...

The only concern I have here is that the comparison with begin() and end() might not be safe if we're already wrongly placed. Boost uses an artificial root node pointer for those for simplicity...